Adding support for more features

phated opened the pull request:
This PR contains commits for:

* Numbers in "e-notation" (2.5e10).
* Support for the union syntax.
* Support for a ListSeparator (`,` or `;`) after a `typedef ...` declaration.

phated commented:
I've also just landed:
* Tests for binary type
* Test for default values in structs
* Test and fix for a number with `-` in the middle.

YanagiEiichi commented on thrift-parser.js:
```diff
@@ -366,6 +410,34 @@ module.exports = (buffer, offset = 0) => {
+  const readUnion = () => {
+    let subject = readKeyword('union');
+    let name = readName();
+    let items = readUnionBlock();
```
Why not use `readStructBlock`?

phated replied:
`readUnionBlock` was needed because it must call `readUnionItem` so it can drop any `option` keyword, which is different behavior than `readStructItem` (see https://github.com/eleme/thrift-parser/pull/6/files#diff-678157f157421c8cb7becf319d0e3b99R431).  I thought about re-using `readStructBlock` but the code would have had to iterate over the `items` and delete each `option` property, which seemed worse than having separated logic.

YanagiEiichi replied:
I think a parser should keep as much original information as possible, even though it's meaningless in some cases.

phated replied:
Typically this results in a warning being emitted from the `thrift` binary because it is invalid syntax.  I have some other work that logs a debug warning if the thrift file has the "required" keyword, but it still deletes the property since it is invalid syntax (whereas "optional" is not invalid).  I'll get some more feedback on my end.

lujjjh replied:
https://github.com/apache/thrift/blob/052abc39c60c54f9901993e76a81036b482ed6ba/compiler/cpp/src/thrift/parse/t_struct.h#L66

Since union members must be optional and unions are treated as `t_struct` as well (according to the [grammar](https://github.com/apache/thrift/blob/330b3f814a33ab8a740fca4da29106bd7e219b1d/compiler/cpp/src/thrift/thrifty.yy#L674)), I agree that union members should include the `option` which is explicitly `optional`.

phated replied:
@lujjjh very interesting.  It also looks like multiple defaults is invalid syntax (and throws), so it seems the `readUnion` logic should remain separate (I'll have to add the singular default logic in another commit).  @YanagiEiichi what do you think?  Should we force each "option" value to "optional"?

YanagiEiichi commented on thrift-parser.js:
```diff
@@ -168,9 +169,51 @@ module.exports = (buffer, offset = 0) => {
 
   const readNumberValue = () => {
     let result = [];
+    if (buffer[offset] === 45) { // -
```
👍

phated commented:
@YanagiEiichi I've updated to keep the original `option` (forcing the default to optional can be done in a compiler). I also renamed the `readStruct*` methods to `readStructLike*` because they are reading things that are Struct-like but not exactly Structs.
